src/test/FreeC/Pass/CompletePatternPassTests.hs

```diff
@@ -0,0 +1,172 @@
+      input <- expectParseTestFuncDecl
+        "f x = case x :: Foobar of {Foo -> Foo ; Bar -> Bar ; Foo -> Foo}"
+      shouldFail $ do
+        _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+        _ <- defineTestCon "Foo" 0 "Foobar"
+        _ <- defineTestCon "Bar" 0 "Foobar"
+        _ <- defineTestVar "x"
+        checkPatternFuncDecl input
+    it "succeeds when all constructors occur exactly once" $ do
+      input <- expectParseTestFuncDecl
+        "f x = case x :: Foobar of {Foo -> Foo ; Bar -> Bar}"
+      shouldSucceed $ do
+        _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+        _ <- defineTestCon "Foo" 0 "Foobar"
+        _ <- defineTestCon "Bar" 0 "Foobar"
+        _ <- defineTestVar "x"
+        checkPatternFuncDecl input
+    it "succeeds when all constructors are given the right number of arguments"
+      $ do
+          input <- expectParseTestFuncDecl
+            "f x = case x :: Foobar of {Foo y -> Foo ; Bar y -> Bar}"
+          shouldSucceed $ do
+            _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+            _ <- defineTestCon "Foo" 1 "Foobar -> Foobar"
+            _ <- defineTestCon "Bar" 0 "Foobar"
+            _ <- defineTestVar "x"
+            checkPatternFuncDecl input
+    it "fails when a constructor of the wrong type occurs" $ do
+      input <- expectParseTestFuncDecl
+        "f x = case x :: Foobar of {Foo -> Foo ; Bar -> Bar ; Evil -> Bar}"
+      shouldFail $ do
+        _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+        _ <- defineTestTypeCon "Evil" 0 ["Evil"]
+        _ <- defineTestCon "Foo" 0 "Foobar"
+        _ <- defineTestCon "Bar" 0 "Foobar"
+        _ <- defineTestCon "Evil" 0 "Evil"
+        _ <- defineTestVar "x"
+        checkPatternFuncDecl input
+  context "Nested and deeper case expressions" $ do
+    it "fails for a faulty case expression inside an if statement" $ do
+      input <- expectParseTestFuncDecl
+        "f x = if b then case x :: Foobar of {Foo -> Foo} else Bar"
+      shouldFail $ do
+        _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+        _ <- defineTestCon "Foo" 0 "Foobar"
+        _ <- defineTestCon "Bar" 0 "Foobar"
+        _ <- defineTestVar "x"
+        _ <- defineTestVar "b"
+        checkPatternFuncDecl input
+    it "fails for a faulty nested case expression" $ do
+      input <-
+        expectParseTestFuncDecl
+          "f x = case x :: Foobar of {Foo -> case x :: Foobar of {Foo -> Foo} ; Bar -> Bar}"
+      shouldFail $ do
+        _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+        _ <- defineTestCon "Foo" 0 "Foobar"
+        _ <- defineTestCon "Bar" 0 "Foobar"
+        _ <- defineTestVar "x"
+        checkPatternFuncDecl input
+    it
+        "fails for a faulty case expression used as another case expression's scrutinee"
+      $ do
+          input <-
+            expectParseTestFuncDecl
+              "f x = case ((case x :: Foobar of {Foo -> x} ) :: Foobar) of {Foo -> Foo ; Bar -> Bar}"
```

Comment:
These lines look like they exceed the 80 character limit. Use string concatenation to break up the lines. Also check other long lines in this file.

```suggestion
    it
        (  "fails for a faulty case expression used as another case"
        ++ "expression's scrutinee"
        )
      $ do
          input <-
            expectParseTestFuncDecl
            $  "f x = case ((case x :: Foobar of {Foo -> x} ) :: Foobar) of {"
            ++ "    Foo -> Foo ;"
            ++ "    Bar -> Bar"
            ++ "  }"
```